File: scripts/dualcam_geom.js
/** 双路货格几何：3D 网格为唯一真值，两路画面都是投影。口径对齐 scripts/dualcam_geom.py。 */

const EPS = 1e-8;
const MIN_DEPTH = 0.05;

function v3(a) {
  return [Number(a[0]), Number(a[1]), Number(a[2])];
}
function add(a, b) { return [a[0] + b[0], a[1] + b[1], a[2] + b[2]]; }
function sub(a, b) { return [a[0] - b[0], a[1] - b[1], a[2] - b[2]]; }
function mul(a, s) { return [a[0] * s, a[1] * s, a[2] * s]; }
function dot(a, b) { return a[0] * b[0] + a[1] * b[1] + a[2] * b[2]; }
function cross(a, b) {
  return [a[1] * b[2] - a[2] * b[1], a[2] * b[0] - a[0] * b[2], a[0] * b[1] - a[1] * b[0]];
}
function norm(a) {
  const n = Math.hypot(a[0], a[1], a[2]) || EPS;
  return mul(a, 1 / n);
}

export function projectPix(p, cam) {
  const C = v3(cam.C);
  const v = sub(v3(p), C);
  const zc = dot(v, cam.fwd);
  if (zc < MIN_DEPTH) return null;
  return [
    cam.cx + cam.f * dot(v, cam.right) / zc,
    cam.cy + cam.f * dot(v, cam.down) / zc,
  ];
}

export function pixelRay(u, v, cam) {
  const d = norm(add(add(mul(cam.right, (u - cam.cx) / cam.f), mul(cam.down, (v - cam.cy) / cam.f)), cam.fwd));
  return { C: v3(cam.C), d };
}

export function wallPlane(corners) {
  const c0 = v3(corners[0]), c1 = v3(corners[1]), c3 = v3(corners[3]);
  return { p0: c0, n: norm(cross(sub(c1, c0), sub(c3, c0))) };
}

export function rayPlane(u, v, cam, p0, n) {
  const { C, d } = pixelRay(u, v, cam);
  const den = dot(d, n);
  if (Math.abs(den) < 1e-8) return null;
  const t = dot(sub(p0, C), n) / den;
  if (t < MIN_DEPTH) return null;
  return add(C, mul(d, t));
}

/** 墙角沿 X 朝巷道平移 inset 米。 */
export function offsetCorners(corners, sign, inset) {
  const dx = -Number(sign) * Number(inset || 0);
  return corners.map((p) => [p[0] + dx, p[1], p[2]]);
}

export function triangulateRays(C1, d1, C2, d2, maxGap = 0.5, attach = 'mid') {
  const w0 = sub(C1, C2);
  const a = dot(d1, d1), b = dot(d1, d2), c = dot(d2, d2);
  const d = dot(d1, w0), e = dot(d2, w0);
  const denom = a * c - b * b;
  if (Math.abs(denom) < 1e-12) return null;
  const t = (b * e - c * d) / denom;
  const s = (a * e - b * d) / denom;
  if (t < MIN_DEPTH || s < MIN_DEPTH) return null;
  const p1 = add(C1, mul(d1, t));
  const p2 = add(C2, mul(d2, s));
  if (Math.hypot(p1[0] - p2[0], p1[1] - p2[1], p1[2] - p2[2]) > maxGap) return null;
  if (attach === 'first') return p1;
  if (attach === 'second') return p2;
  return mul(add(p1, p2), 0.5);
}

export function triangulatePixels(u1, v1, cam1, u2, v2, cam2, attach = 'mid') {
  const a = pixelRay(u1, v1, cam1);
  const b = pixelRay(u2, v2, cam2);
  return triangulateRays(a.C, a.d, b.C, b.d, 0.5, attach);
}

/** 默认射线∩开口平面。stereo 时本路跟鼠标。 */
export function dragVertex(u, v, cam, otherCam, current, p0, n, stereo = false) {
  if (stereo) {
    const uvO = projectPix(current, otherCam);
    if (uvO) {
      const p = triangulatePixels(u, v, cam, uvO[0], uvO[1], otherCam, 'first');
      if (p) return p;
    }
  }
  return rayPlane(u, v, cam, p0, n);
}

export function makeGridVertices(corners, rows, cols) {
  const c0 = v3(corners[0]), c1 = v3(corners[1]), c2 = v3(corners[2]), c3 = v3(corners[3]);
  const out = [];
  for (let r = 0; r <= rows; r++) {
    const ty = r / rows;
    for (let c = 0; c <= cols; c++) {
      const tz = c / cols;
      out.push(add(add(
        add(mul(c0, (1 - ty) * (1 - tz)), mul(c1, (1 - ty) * tz)),
        mul(c2, ty * tz),
      ), mul(c3, ty * (1 - tz))));
    }
  }
  return out;
}

export function vertIndex(rows, cols, r, c) {
  return r * (cols + 1) + c;
}

export function meshCells(mesh) {
  const rows = mesh.rows, cols = mesh.cols, verts = mesh.vertices;
  const cells = [];
  for (let i = 0; i < rows; i++) {
    for (let j = 0; j < cols; j++) {
      cells.push({
        row: i,
        col: j,
        box_id: `r${i}c${j}`,
        corners: [
          verts[vertIndex(rows, cols, i, j)],
          verts[vertIndex(rows, cols, i, j + 1)],
          verts[vertIndex(rows, cols, i + 1, j + 1)],
          verts[vertIndex(rows, cols, i + 1, j)],
        ],
      });
    }
  }
  return cells;
}

export function wallById(solved, wallId) {
  return ((solved && solved.walls) || []).find((w) => w.wall_id === wallId) || null;
}

export function signedWallDist(p, wall) {
  const p0 = v3(wall.corners[0]);
  const nx = Number(wall.sign) < 0 ? 1 : -1;
  return (p[0] - p0[0]) * nx;
}

export function contactSlots(p, meshes, solved, contactM = 0.10) {
  if (!p) return [];
  const hits = [];
  for (const mesh of meshes || []) {
    const wall = wallById(solved, mesh.wall_id);
    if (!wall) continue;
    const d = signedWallDist(p, wall);
    if (d > contactM) continue;
    const yz = [p[1], p[2]];
    for (const cell of meshCells(mesh)) {
      const poly = cell.corners.map((c) => [c[1], c[2]]);
      if (pointInPolygon(yz, poly)) {
        hits.push({
          wall_id: mesh.wall_id,
          box_id: cell.box_id,
          row: cell.row,
          col: cell.col,
          d,
        });
      }
    }
  }
  return hits;
}

export function pointInPolygon(point, polygon) {
  if (!polygon || polygon.length < 3) return false;
  const [x, y] = point;
  let inside = false;
  for (let i = 0, j = polygon.length - 1; i < polygon.length; j = i++) {
    const xi = polygon[i][0], yi = polygon[i][1];
    const xj = polygon[j][0], yj = polygon[j][1];
    const hit = (yi > y) !== (yj > y) && x < ((xj - xi) * (y - yi)) / (yj - yi || 1e-12) + xi;
    if (hit) inside = !inside;
  }
  return inside;
}

/** 3D 四角投影到某路；缺深度则整格丢掉。 */
export function projectCorners(corners, cam) {
  const pts = [];
  for (const p of corners) {
    const uv = projectPix(p, cam);
    if (!uv) return null;
    pts.push(uv);
  }
  return pts;
}

File: scripts/dualcam_geom.py
"""双路货格几何：3D 网格为唯一真值，两路画面都是投影。

世界系与 solve_scene 相同：X 横跨巷道、Y 向上、Z 沿巷道。
墙角顺序与标注一致：①顶远 ②顶近 ③底近 ④底远（远近相对左路）。
"""

from __future__ import annotations

from typing import Any

import numpy as np

EPS = 1e-8
MIN_DEPTH = 0.05


def _v(a: Any) -> np.ndarray:
    return np.asarray(a, dtype=float).reshape(-1)


def project_pix(p: Any, cam: dict) -> list[float] | None:
    """世界点 → 像素。相机背后返回 None。"""
    C = _v(cam["C"])
    fwd, right, down = _v(cam["fwd"]), _v(cam["right"]), _v(cam["down"])
    f, cx, cy = float(cam["f"]), float(cam["cx"]), float(cam["cy"])
    v = _v(p) - C
    zc = float(v @ fwd)
    if zc < MIN_DEPTH:
        return None
    return [cx + f * float(v @ right) / zc, cy + f * float(v @ down) / zc]


def pixel_ray(u: float, v: float, cam: dict) -> tuple[np.ndarray, np.ndarray]:
    """像素 → 世界系射线 (原点, 单位方向)。"""
    C = _v(cam["C"])
    fwd, right, down = _v(cam["fwd"]), _v(cam["right"]), _v(cam["down"])
    f, cx, cy = float(cam["f"]), float(cam["cx"]), float(cam["cy"])
    d = ((u - cx) / f) * right + ((v - cy) / f) * down + fwd
    n = float(np.linalg.norm(d))
    return C, d / max(n, EPS)


def wall_plane(corners: Any) -> tuple[np.ndarray, np.ndarray]:
    """墙四角 ①②③④ → 平面 (点, 单位法向)。"""
    c = np.asarray(corners, dtype=float)
    n = np.cross(c[1] - c[0], c[3] - c[0])
    n = n / max(float(np.linalg.norm(n)), EPS)
    return c[0], n


def ray_plane(u: float, v: float, cam: dict, p0: Any, n: Any) -> list[float] | None:
    """像素射线 ∩ 平面。不相交或朝后返回 None。"""
    C, d = pixel_ray(u, v, cam)
    p0, n = _v(p0), _v(n)
    den = float(d @ n)
    if abs(den) < 1e-8:
        return None
    t = float((p0 - C) @ n) / den
    if t < MIN_DEPTH:
        return None
    return (C + t * d).tolist()


def offset_corners(corners: Any, sign: int, inset: float) -> list[list[float]]:
    """墙角沿 X 朝巷道平移 inset 米，得到货格开口所在平面。"""
    dx = -float(sign) * float(inset)
    return [[float(p[0]) + dx, float(p[1]), float(p[2])] for p in corners]


def triangulate_rays(
    c1: Any, d1: Any, c2: Any, d2: Any, max_gap: float = 0.5, attach: str = "mid",
) -> list[float] | None:
    """两射线最近点。attach=first 则落在第一条射线上（拖哪路跟哪路）。"""
    c1, d1, c2, d2 = _v(c1), _v(d1), _v(c2), _v(d2)
    w0 = c1 - c2
    a, b, c = float(d1 @ d1), float(d1 @ d2), float(d2 @ d2)
    d, e = float(d1 @ w0), float(d2 @ w0)
    denom = a * c - b * b
    if abs(denom) < 1e-12:
        return None
    t = (b * e - c * d) / denom
    s = (a * e - b * d) / denom
    if t < MIN_DEPTH or s < MIN_DEPTH:
        return None
    p1, p2 = c1 + t * d1, c2 + s * d2
    if float(np.linalg.norm(p1 - p2)) > max_gap:
        return None
    if attach == "first":
        return p1.tolist()
    if attach == "second":
        return p2.tolist()
    return (0.5 * (p1 + p2)).tolist()


def triangulate_pixels(
    u1: float, v1: float, cam1: dict, u2: float, v2: float, cam2: dict,
    attach: str = "mid",
) -> list[float] | None:
    c1, d1 = pixel_ray(u1, v1, cam1)
    c2, d2 = pixel_ray(u2, v2, cam2)
    return triangulate_rays(c1, d1, c2, d2, attach=attach)


def drag_vertex(
    u: float, v: float, cam: dict, other_cam: dict, current: Any, p0: Any, n: Any,
    stereo: bool = False,
) -> list[float] | None:
    """默认射线∩开口平面。stereo=True 时本路跟鼠标、用另一路像素三角化。"""
    if stereo:
        uv_o = project_pix(current, other_cam)
        if uv_o is not None:
            p = triangulate_pixels(u, v, cam, uv_o[0], uv_o[1], other_cam, attach="first")
            if p is not None:
                return p
    return ray_plane(u, v, cam, p0, n)


def make_grid_vertices(corners: Any, rows: int, cols: int) -> list[list[float]]:
    """在墙矩形上均匀切 rows×cols 格，顶点 (rows+1)*(cols+1)，行主序。

    r=0 顶、c=0 左路远端。双线性即矩形插值（墙在 3D 是矩形）。
    """
    if rows < 1 or cols < 1:
        raise ValueError("rows/cols 至少为 1")
    c0, c1, c2, c3 = [_v(p) for p in corners]
    out: list[list[float]] = []
    for r in range(rows + 1):
        ty = r / rows
        for c in range(cols + 1):
            tz = c / cols
            p = (
                (1 - ty) * (1 - tz) * c0
                + (1 - ty) * tz * c1
                + ty * tz * c2
                + ty * (1 - tz) * c3
            )
            out.append([float(x) for x in p])
    return out


def vert_index(rows: int, cols: int, r: int, c: int) -> int:
    return r * (cols + 1) + c


def mesh_cells(mesh: dict) -> list[dict]:
    """网格 → 单元格列表，每格四角仍是 ①顶远 ②顶近 ③底近 ④底远。"""
    rows, cols = int(mesh["rows"]), int(mesh["cols"])
    verts = mesh["vertices"]
    cells = []
    for i in range(rows):
        for j in range(cols):
            corners = [
                verts[vert_index(rows, cols, i, j)],
                verts[vert_index(rows, cols, i, j + 1)],
                verts[vert_index(rows, cols, i + 1, j + 1)],
                verts[vert_index(rows, cols, i + 1, j)],
            ]
            cells.append({
                "row": i,
                "col": j,
                "box_id": f"r{i}c{j}",
                "corners": corners,
            })
    return cells


def wall_by_id(solved: dict, wall_id: int) -> dict | None:
    for w in solved.get("walls") or []:
        if w.get("wall_id") == wall_id:
            return w
    return None


def _point_in_poly(pt: Any, poly: Any) -> bool:
    x, y = float(pt[0]), float(pt[1])
    inside = False
    n = len(poly)
    j = n - 1
    for i in range(n):
        xi, yi = float(poly[i][0]), float(poly[i][1])
        xj, yj = float(poly[j][0]), float(poly[j][1])
        if (yi > y) != (yj > y) and x < (xj - xi) * (y - yi) / (yj - yi + 1e-12) + xi:
            inside = not inside
        j = i
    return inside


def signed_wall_dist(p: Any, wall: dict) -> float:
    """巷道内侧为正：人在通道里 >0，贴墙/伸进 ≤0。"""
    p0 = _v(wall["corners"][0])
    inward = np.array([1.0 if int(wall.get("sign", -1)) < 0 else -1.0, 0.0, 0.0])
    return float((_v(p) - p0) @ inward)


def contact_slots(
    p: Any, meshes: list[dict], solved: dict, contact_m: float = 0.10,
) -> list[dict]:
    """腕点接触墙面且落在货格 (Y,Z) 多边形内 → 该货位报警。"""
    if p is None:
        return []
    hits = []
    for mesh in meshes or []:
        wall = wall_by_id(solved, mesh.get("wall_id"))
        if not wall:
            continue
        d = signed_wall_dist(p, wall)
        if d > contact_m:
            continue
        yz = [_v(p)[1], _v(p)[2]]
        for cell in mesh_cells(mesh):
            poly = [[c[1], c[2]] for c in cell["corners"]]
            if _point_in_poly(yz, poly):
                hits.append({
                    "wall_id": mesh["wall_id"],
                    "box_id": cell["box_id"],
                    "row": cell["row"],
                    "col": cell["col"],
                    "d": round(d, 3),
                })
    return hits
